### PR TITLE
Add worker script explanation to COOP-COEP article

### PR DESCRIPTION
**Pull request by agektmr**
This pull request updates https://web.dev/cross-origin-isolation-guide/ and https://web.dev/coop-coep/ to better explain how to opt-in to cross-origin isolation when worker scripts are used.
cc: @ArthurSonzogni

### CODE REVIEW

**Review by ArthurSonzogni — COMMENTED** [review]
Nice!
Here are some comments below:

File: src/site/content/en/blog/coop-coep/index.md [hunk 1]
```diff
@@ -204,11 +204,11 @@ terminate the origin trial until this issue is safely resolved.
 
 {% endAside %}
 
-A browsing context group is a group of tabs, windows or iframes which share the
-same context. For example, if a website (`https://a.example`) opens a popup
-window (`https://b.example`), the opener window and the popup window share the
-same browsing context and they have access to each other via DOM APIs such as
-`window.opener`.
+A browsing context group is a group of tabs, windows, iframes or worker scripts
```
[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:207]: I added an explanation of "browsing context group" because the term is used in other places in this document. I think precise understanding of what is browsing context group is unnecessary or distracting for general readers. I'd like to phrase this in a way it is not too detailed, but not telling a lie.

Do you have any suggestions?

File: src/site/content/en/secure/cross-origin-isolation-guide/index.md [hunk 2]
```diff
@@ -178,12 +178,12 @@ cross-origin resources:
    `<img src="example.jpg" crossorigin>`).
 4. If you want to use powerful features such as `SharedArrayBuffer` inside a
```
[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/secure/cross-origin-isolation-guide/index.md:179]: Good catch!

File: src/site/content/en/blog/coop-coep/index.md [hunk 3]
```diff
@@ -260,19 +261,25 @@ same-origin](/same-site-same-origin).
...
 `Cross-Origin-Embedder-Policy-Report-Only` header to examine whether the policy
 actually works. You will receive reports without blocking embedded content.
+
 Recursively apply this to all documents. For information on the Report-Only HTTP
 header, see [Observe issues using the Reporting
 API](#observe-issues-using-the-reporting-api).
 
+For iframes and worker scripts, don't forget to send a
+`Cross-Origin-Embedder-Policy-Report-Only` header with the same value as the
+parent document in addition to a CORP header.
```
[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md]: The previous paragraph included "recursively apply" and I meant to follow that paragraph, but I think it's easier just to merge them.

File: src/site/content/en/secure/cross-origin-isolation-guide/index.md [hunk 4]
```diff
@@ -178,12 +178,12 @@ cross-origin resources:
    `<img src="example.jpg" crossorigin>`).
```
[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/secure/cross-origin-isolation-guide/index.md]: Rephrased.

File: src/site/content/en/blog/coop-coep/index.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArthurSonzogni) [reply src/site/content/en/blog/coop-coep/index.md:209]: Maybe you can just say.

A browsing context group is a set of window that can reference each other. For instance, the top-level document and its child documents embedded via <iframe>, or popup opened using window.open.

[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:209]: Can I say:

A browsing context group is a set of window that can reference each other. For instance, the top-level document and its child documents embedded via <iframe>, **worker scripts spawned within the document** or popup opened using window.open.

[the remark this answers is not in the document]
Comment (ArthurSonzogni) [reply src/site/content/en/blog/coop-coep/index.md:209]: I don't think so ;-)

DedicatedWorker can interact only with its creator, so I won't feel bad saying they "belong" to the same group as their creator, in an extended sense. However, SharedWorker and ServiceWorker can interact with every same-origin document, no matter their browsing context group. So, I would feel bad saying a SharedWorker/ServiceWorker is associated with a defined browsing context group.

Per the specification, I don't believe we can say a Worker is a "browsing context". They talk only about a window/document like object.

However, even if we don't really define browsing context group for workers, there are several COEP rules to know about:
- If a document has COEP, it's DedicatedWorker must have COEP.
- If a document has COEP, its SharedWorker must have COEP.
- ServiceWorker are free to live independently. However, avoiding a mismatch of COEP matter, for process allocation.
- To avoid leaking resources requested without COEP toward a context with COEP, we make responses sent using CacheStorage to execute the CORP check again on the response with the outgoing context, like they would have done during a network fetch.

[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:209]: I will omit the `worker scripts spawned within the document` and use your suggested phrase as is.

```
A browsing context group is a set of window that can reference each other. For instance, the top-level document and its child documents embedded via <iframe>, or popup opened using window.open.
```

Could you rephrase this for me to better understand what it means?

```
ServiceWorker are free to live independently. However, avoiding a mismatch of COEP matter, for process allocation.
```

I wonder how to use ServiceWorker on cross-origin isolated environment is another tip developers might look for.

[the remark this answers is not in the document]
Comment (ArthurSonzogni) [reply src/site/content/en/blog/coop-coep/index.md:209]: We should ask developers to add COEP in both the Document and its ServiceWorker There are two reasons:
- The ServiceWorker process will be allowed to stay in the same process as their document. It saves resources and improves performances.
- A resource successfully received by the SW will always be allowed to be transferred to the document via the CacheStorage.

Contrary to DedicatedWorker, it is possible to use a ServiceWorker with a mismatching COEP. This is somehow a problem, because developers might not understand by themselves they should add COEP in ServiceWorker.

[the remark this answers is not in the document]
Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:209]: Sounds like more complex than I have expected.
I'll resolve this issue for now just to cover Web Workers and will add Service Worker section later on.

**Comment by agektmr** [comment]
@ArthurSonzogni Please approve if the current changes look good

File: src/site/content/en/blog/coop-coep/index.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArthurSonzogni) [reply src/site/content/en/blog/coop-coep/index.md]: SGTM!

File: src/site/content/en/blog/coop-coep/index.md [hunk 5]
```diff
@@ -204,11 +206,11 @@ terminate the origin trial until this issue is safely resolved.
...
-A browsing context group is a group of tabs, windows or iframes which share the
-same context. For example, if a website (`https://a.example`) opens a popup
-window (`https://b.example`), the opener window and the popup window share the
-same browsing context and they have access to each other via DOM APIs such as
-`window.opener`.
+A browsing context group is a set of window that can reference each other. For
+example, the top-level document and its child documents embedded via `<iframe>`.
+If a website (`https://a.example`) opens a popup window (`https://b.example`),
+the opener window and the popup window share the same browsing context and they
+have access to each other via DOM APIs such as `window.opener`.
```
Comment (heyawhite) [inline src/site/content/en/blog/coop-coep/index.md:213]:
```suggestion
A browsing context group is a set of windows that can reference each other. For
example, a top-level document and its child documents embedded via `<iframe>`.
If a website (`https://a.example`) opens a popup window (`https://b.example`),
the opener window and the popup window share the same browsing context, therefore
they have access to each other via DOM APIs such as `window.opener`.
```

File: src/site/content/en/blog/coop-coep/index.md [hunk 6]
```diff
@@ -237,16 +239,17 @@ Here is what you need to do depending on the nature of the resource:
     * Use the `crossorigin` attribute in the loading HTML tag if the resource is
       served with CORS. (For example, `<img src="***" crossorigin>`.)
     * Ask the owner of the resource to support either CORS or CORP.
-* For iframes, use CORP and COEP headers as follows:
-  `Cross-Origin-Resource-Policy: same-origin` (or `same-site`, `cross-origin`
-  depending on the context) and `Cross-Origin-Embedder-Policy: require-corp`.
+* For iframes and worker scripts, set the `Cross-Origin-Resource-Policy:
```
Comment (ArthurSonzogni) [inline src/site/content/en/blog/coop-coep/index.md:242]: Setting CORP is needed only for cross-origin iframe. Not really useful for workers. They are always same-origin.
So, I would remove the "and worker scripts".

COEP on the other site is needed or useful for every environments created by the document. You moved the section later. This is discussed separately, which works.

**Review by ArthurSonzogni — COMMENTED** [review]
Thanks! LGTM % removing the sentence about CORP and workers.

File: src/site/content/en/blog/coop-coep/index.md (hunk 6 above)
Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:242]: You can only spawn a same-origin worker at the first level, but can load a cross-origin script via that worker if you use `importScripts`. In that case, you have to use `CORP: cross-origin` header. You can see it working at https://cross-origin-isolation.glitch.me/?coep=require-corp&

Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:242]: Hmm, this contradicts 🤔
https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers#spawning_subworkers

Comment (ArthurSonzogni) [reply src/site/content/en/blog/coop-coep/index.md:242]: I understood "worker script" as the "worker's main response".

This is right worker can then use fetch/importScript to load any resources and those resources are subject to CORP checks. Yes, you can nested DedicatedWorker.

Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:242]: In that case, how about phrasing the line like this:

```
For iframes and cross-origin worker scripts loaded via `importScripts`, set the `Cross-Origin-Resource-Policy:
```

Comment (ArthurSonzogni) [reply src/site/content/en/blog/coop-coep/index.md:242]: This feels weird, because it would mean this applies exclusively to those two resources, which is wrong; it applies to every resources.
Also, I feel weird putting on the same level the main resources (the iframe's response) and a subresource (an external script loaded from the worker).

Maybe you can discuss about all subresources, indepedently of the context (window or worker), and give some examples of subresources?
---

- From a `COEP:require-corp` document or a `COEP:require-corp` worker, cross-origin subresources loaded without CORS must set `Cross-Origin-Resource-Policy: cross-origin` header to opt-in being embedded. For instance, this applies to: `<script>`, `importScript`, `<link>`, `<video>`, `<iframe>`, etc...

Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:242]: That's a great idea. I'll create a pull request so we can continue the discussion there.

Comment (agektmr) [reply src/site/content/en/blog/coop-coep/index.md:242]: https://github.com/GoogleChrome/web.dev/pull/8239
